### ISSUE TITLE
Exclude rule config is not honored

### ISSUE THREAD

**Issue by puetsch** [sqlfluff/sqlfluff#2656]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I added config to exclude rule L029 but it's still being processed.

### Expected Behaviour

Rule L029 is disabled

### Observed Behaviour

Rule L029 is being triggered although the verbose output shows it's deactivated.

>     exclude_rules:      L009,L011,L019,L031,L034,L029
vs.
> L:   1 | P:  13 | L029 | Keywords should not be used as identifiers.

```
❯ sqlfluff lint -vv ../test3.sql
==== sqlfluff ====
sqlfluff:               0.10.0 python:                  3.7.6
implementation:        cpython dialect:              postgres
verbosity:                   2 templater:               jinja

== Raw Config:
core:                                   
    dialect:            postgres            
    disable_noqa:       False               
    encoding:           autodetect          
    exclude_rules:      L009,L011,L019,L031,L034,L029
    fix_even_unparsable:False               
    ignore:             []                  
    ignore_templated_areas:True                
    nocolor:            False               
    output_line_length: 160                 
    recurse:            True                
    runaway_limit:      10                  
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja               
    verbose:            2                   
indentation:                            
    indented_ctes:      False               
    indented_joins:     False               
    indented_using_on:  True                
    template_blocks_indent:True                
rules:                                  
    allow_scalar:       True                
    comma_style:        trailing            
    indent_unit:        space               
    max_line_length:    160                 
    single_table_references:consistent          
    tab_space_size:     4                   
    unquoted_identifiers_policy:all                 
    L003:                                   
        lint_templated_tokens:True                
    L007:                                   
        operator_new_lines: after               
    L010:                                   
        capitalisation_policy:consistent          
    L011:                                   
        aliasing:           explicit            
    L012:                                   
        aliasing:           explicit            
    L014:                                   
        extended_capitalisation_policy:consistent          
    L016:                                   
        ignore_comment_clauses:False               
        ignore_comment_lines:False               
    L026:                                   
        force_enable:       False               
    L028:                                   
        force_enable:       False               
    L029:                                   
        unquoted_identifiers_policy:aliases             
    L030:                                   
        capitalisation_policy:consistent          
    L038:                                   
        select_clause_trailing_comma:forbid              
    L040:                                   
        capitalisation_policy:consistent          
    L042:                                   
        forbid_subquery_in: join                
    L047:                                   
        prefer_count_0:     False               
        prefer_count_1:     False               
    L052:                                   
        multiline_newline:  False               
        require_final_semicolon:False               
    L054:                                   
        group_by_and_order_by_style:consistent          
    L057:                                   
        additional_allowed_characters:""                  
        allow_space_in_identifier:False               
        quoted_identifiers_policy:all                 
        unquoted_identifiers_policy:all                 
    L059:                                   
        prefer_quoted_identifiers:False               
    L062:                                   
templater:                              
    unwrap_wrapped_queries:True                
    jinja:                                  
        apply_dbt_builtins: True                
        macros:                                 
            dbt_config:         {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
            dbt_is_incremental: {% macro is_incremental() %}True{% endmacro %}
            dbt_ref:            {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
            dbt_source:         {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
            dbt_var:            {% macro var(variable) %}item{% endmacro %}

==== readout ====

=== [ path: ../test3.sql ] ===

== [../test3.sql] LINTING                                                                               
== [../test3.sql] FAIL                                                                                  
L:   1 | P:  13 | L029 | Keywords should not be used as identifiers.
==== summary ====
files:             1 violations:        1
clean files:       0 unclean files:     1
avg per file:   1.00 unclean rate:   100%
status:         FAIL
All Finished 📜 🎉!

### How to reproduce

`sqlfluff lint -vv ../test3.sql`

using

```sql
select a as cost
from b

```

### Dialect

redshift

### Version

`sqlfluff:               0.10.0 python:                  3.7.6`

### Configuration

```
[sqlfluff]
verbose = 0
nocolor = False
dialect = postgres
templater = jinja
rules = None
exclude_rules = L009,L011,L019,L031,L034,L029
recurse = 0
output_line_length = 160
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 160
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = consistent

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = consistent

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
The only way I can repeat this is by having a DIFFERENT `.sqlfluff` file in the `../` directory which does not exclude L029.

i.e. you have the following

```
# Main config for this project, which excludes L029
./.sqlfluff
# Additional config which does NOT exclude L029 and is used for test3.sql over the above file.
../.sqlfluff
../test3.sql
```

Could that be the case here?

i.e. The verbose output shows the current config, for the current directory, but then you ask it to look at a different directory and that directly has some overridden config. I admit it's confusing from the verbose output.

**Comment by puetsch** [comment]
That is so!

When I remove the file in `..` and only keep the config in the current dir, it works as expected.

**Comment by tunetheweb** [comment]
Or if you remove `exclude_rules` from the file in `..` then it won't override them and will use the base config from this directory for that parameter.

Or alternatively you can force it to use the current config:

```
sqlfluff lint ../test3.sql --config .sqlfluff 
```

This is the way SQLFluff works so sounds like it's working as intended. So closing.

**Comment by tunetheweb** [comment]
It's not quite as simple as that.

The verbose output is outputing the config from the current directory. It is correct to output that in my humble opinion as that's the base config that SQLFLuff is initialised with. Then it overwrites it for the SQL in other directories if a more specific `.sqlfluff` config exists for any SQL infiles/directories it's being as to lint [as given by the nesting rules](https://docs.sqlfluff.com/en/stable/configuration.html#nesting).

Put it another way if you had this:

```
.sqlfluff
test1.sql
test2/.sqlfluff
test2/test2.sql
test3/.sqlfluff
test3/test3.sql
test4/test4.sql
```

And you ran `sqlfluff lint ./` the following would happen:

1. The base `.sqlfluff` to be read and used to lint` test1.sql`
2. Any differences in `test2/.sqlfluff` are applied on top of `.sqlfluff` and then used to lint `test2/test2.sql`
3. Any differences in `test3/.sqlfluff` are applied on top of `.sqlfluff` and then used to line `test3/test3.sql`
4. The base `.sqlfluff` is used to lint `test4.sql`

So, back to your example, it's not true that it used `../.sqlfluff` for linting your file (and so that's what should have been printed out in preference to `./.sqlfluff` contents). It actually used a combination of `./sqlfluff` and `../.sqlfluff` (though it's possible your more specific `.sqlfluff` file completely overwrote all the settings in original base one if it has the same config lines). The core config was loaded from the current directory (and that's what is outputed with `-vv`) and THEN it looks at the files, and see that config is subsequently overridden. This can be seen by not including `exclude_rules = None` in your `../.sqlfluff` file and you'll see it does retain this setting from the `.sqlfluff` file that was originally loaded, when it's not specifically being override to `None`.

We probably could (maybe should?) make it more obvious that more config is loaded on top of that base config, though that has it's only logging challenges as could be just as confusing. In my example above, the `test3/.sqlfluff` config isn't applied on top of `test2/.sqlfluff` but that is reset and it's actually applied on top of `.sqlfluff`.

So having debug log like this:

```
Loading .sqlfluff
... list all config options...
Linting test1.sql
Linting test2 directory
Loading test2/.sqlfluff and override .sqlfluff for same keys
... list all config options...
Linting test2.sql
Reseting config to original base config ./.sqlfluff
... list all config options...
Linting test3 directory
Loading test3/.sqlfluff and override .sqlfluff for same keys
... list all config options...
Linting test3.sql
Reseting config to original base config ./.sqlfluff
... list all config options...
Linting test4 directory
Linting test4.sql
```

Is probably what we really want, but that's very confusing too.

The idea, of all this nesting btw, is you set top-level SQLFLuff rules in your top level folder, and then only override more specific config for other folders. So, if you were using mixed dialect your could have this:

```
.sqlfluff
mysql/.sqlfluff
mysql/test2.sql
postgres/.sqlfluff
postgres/test3.sql
```

You could have all your base config in the top level `.sqlfluff` file, and then only override the `dialect` in each of the subfolder `.sqlfluff` files so those are both very light 2 line `.sqfluff` files.

At the moment SQLFluff errors on the side of not being overly verbose here (even in verbose mode!) and just lists the first config. I agree that can be confusing (as it was for you), but can also be confusing, the other way as per above example. I'm noit sure what the right answer he is (other than avoid running SQLFluff with so many `.sqlfluff` files where possible!).

I guess the single file example you had is doubly confusing, as it may seem obvious it should just print that config, and not waste it's time, with the `.sqfluff` config in the current directory, but as per above it's not quite as a simple as that. Plus SQLFluff is also built in mind with more than linting a single file (as per above example), so has to consider how to handle that.

Anyway, the reason I close the issue is your question was answered - we know why the `exclude_rule` was "not honoured" - it was overwritten by later config and that "was honoured" in it's place. As it should have been.

If you want to suggest changing the logic here, or adding additional logging, then can you change the issue title and description? Though to be honest it might be cleaner to open a new one and refer to this one instead so we can start from a fresh slate with the new ask now we've clarified why your original ask wasn't the right question.

Does that all make sense?

**Comment by puetsch** [comment]
Thank you so much for elaborating on this. I def learnt a lot about how sqlfluff handles the config files.

I *did* expect that the outputted config would 1:1 relate to the config that underlies the subsequent lint results.

I understand now that this is a little more involved, and potentially each file (folder) could even have a changed config file.

The cleanest solution might be to also have `-vv` output any additional config file it picks up "along the way"—like you outlined in your second to last snippet.

I'll open a new issue for this — so in case someone else gets equally confused they can comment or upvote it.

**Comment by Gregory108** [comment]
Hi! Came here after a misunderstanding caused me to mis-configure pre-commit hooks.
As I observe from practice and (seem to) understand from the discussion above, passing arguments to CLI/pre-commit does not append, but overrides each configuration parameter. Is that correct?
E.g. `.sqlfluff` has `exclude_rules: L032`, but when pre-commit `args: [--exclude-rules, "L029"]`, the L032 is applied.

I think this is okay (or, maybe even the right way!). But I wish warning that argument override config would make its way into documentations ([here](https://docs.sqlfluff.com/en/stable/cli.html))
